### ISSUE TITLE
Rule Change: preserve-caught-error allow extra data in cause

### ISSUE THREAD

**Issue by andrei9669** [eslint/eslint#20108]
### What rule do you want to change?

preserve-caught-error

### What change do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
} catch (e) {
    throw new Error("Error 2", { cause: { Error: e, some: "data" } });
}
```

### What does the rule currently do for this code?

It gives the following warning: `ESLint: The symptom error is being thrown with an incorrect 'cause'. (preserve-caught-error)`

### What will the rule do after it's changed?

no warning when the error is present in cause object.

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

in my use case, I would also want to pass additional data to the cause, for example, function parameters what could have caused the error or some other meta data.

Ideally, the rule should not give errors as long as the error param is included in the cause object directly.

**Comment by mdjermanovic** [comment]
I believe the convention is to pass caught error as `cause` when rethrowing, for proper chaining, and that's what environments/tools expect.

For example:

```js
function a() {
    throw new Error("Something unexpected occured");
}

function b() {
    try {
        a();
    } catch (e) {
        throw new Error("Error while doing some work", { cause: e });
    }
}

function c() {
    try {
        b();
    } catch (e) {
        throw new Error("Error while doing another work", {
            cause: e,
        });
    }
}

c();

```

In Chrome, this outputs the entire chain:

<img width="490" height="520" alt="Image" src="https://github.com/user-attachments/assets/0c49752a-e4ce-45f9-9b3b-23f95078127d" />

Now, if we do `{ cause: { Error: e, some: "data" } }` as in the original post:

```js
function a() {
    throw new Error("Something unexpected occured");
}

function b() {
    try {
        a();
    } catch (e) {
        throw new Error("Error while doing some work", { cause: e });
    }
}

function c() {
    try {
        b();
    } catch (e) {
        throw new Error("Error while doing another work", { 
            cause: {
                Error: e,
                some: "data" 
            }
        });
    }
}

c();

```

It outputs only the last error:

<img width="496" height="426" alt="Image" src="https://github.com/user-attachments/assets/e0867e57-9e68-4711-9e99-346ac6594638" />

So, I think this rule enforcing `{ cause: e }` is correct behavior, and shouldn't be modified to allow any object with `e` inside.

@eslint/eslint-team thoughts?

**Comment by JoshuaKGoldberg** [comment]
Tricky. Technically, `cause` _can_ be anything. From https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause#description:

> The value of `cause` can be of any type. You should not make assumptions that the error you caught has an `Error` as its cause, in the same way that you cannot be sure the variable bound in the `catch` statement is an `Error` either. The "Providing structured data as the error cause" example below shows a case where a non-error is deliberately provided as the cause.

However, I agree with mdjermanovic that the _convention_ is to use an `Error` as `cause`. Perhaps allowing an object value should be an opt-in option? That value would need to have the original error somewhere in its properties.

**Comment by fasttime** [comment]
While technically the `cause` property can be any value, the convention is to set it to an `Error` object so it can be recursively discovered by logging tools. For example, Node.js will collapse identical stack trace lines and show the original error when `cause` is used like this:

```js
const error1 = new Error("Error 1");
const error2 = new Error("Error 2", { cause: error1 });
const error3 = new Error("Error 3", { cause: error2 });
const error4 = new Error("Error 4", { cause: error3 });
console.log(error4);
```

prints:

```text
Error: Error 4
    at Object.<anonymous> (.../test.js:16:16)
    at Module._compile (node:internal/modules/cjs/loader:1706:14)
    ... 6 lines matching cause stack trace ...
    at node:internal/main/run_main_module:36:49 {
  [cause]: Error: Error 3
      at Object.<anonymous> (.../test.js:15:16)
      at Module._compile (node:internal/modules/cjs/loader:1706:14)
      ... 6 lines matching cause stack trace ...
      at node:internal/main/run_main_module:36:49 {
    [cause]: Error: Error 2
        at Object.<anonymous> (.../test.js:14:16)
        at Module._compile (node:internal/modules/cjs/loader:1706:14)
        ... 6 lines matching cause stack trace ...
        at node:internal/main/run_main_module:36:49 {
      [cause]: Error: Error 1
          at Object.<anonymous> (.../test.js:13:16)
          at Module._compile (node:internal/modules/cjs/loader:1706:14)
          at Object..js (node:internal/modules/cjs/loader:1839:10)
          at Module.load (node:internal/modules/cjs/loader:1441:32)
          at Function._load (node:internal/modules/cjs/loader:1263:12)
          at TracingChannel.traceSync (node:diagnostics_channel:322:14)
          at wrapModuleLoad (node:internal/modules/cjs/loader:237:24)
          at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:171:5)
          at node:internal/main/run_main_module:36:49
    }
  }
}
```

This doesn't work when the error object is wrapped as a property of another object:

```js
const error1 = new Error("Error 1");
const error2 = new Error("Error 2", { cause: { error: error1 } });
const error3 = new Error("Error 3", { cause: { error: error2 } });
const error4 = new Error("Error 4", { cause: { error: error3 } });
console.log(error4);
```

prints:

```text
Error: Error 4
    at Object.<anonymous> (.../test.js:16:16)
    at Module._compile (node:internal/modules/cjs/loader:1706:14)
    at Object..js (node:internal/modules/cjs/loader:1839:10)
    at Module.load (node:internal/modules/cjs/loader:1441:32)
    at Function._load (node:internal/modules/cjs/loader:1263:12)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:237:24)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:171:5)
    at node:internal/main/run_main_module:36:49 {
  [cause]: {
    error: Error: Error 3
        at Object.<anonymous> (.../test.js:15:16)
        at Module._compile (node:internal/modules/cjs/loader:1706:14)
        at Object..js (node:internal/modules/cjs/loader:1839:10)
        at Module.load (node:internal/modules/cjs/loader:1441:32)
        at Function._load (node:internal/modules/cjs/loader:1263:12)
        at TracingChannel.traceSync (node:diagnostics_channel:322:14)
        at wrapModuleLoad (node:internal/modules/cjs/loader:237:24)
        at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:171:5)
        at node:internal/main/run_main_module:36:49 {
      [cause]: [Object]
    }
  }
}
```

I'd recommend to refactor your project so it aligns to the standard usage. If that's not possible and you want ESLint to enforce your pattern, you could create a custom rule for that.

**Comment by fregante** [comment]
`cause` can be anything because errors can be anything. I wouldn't set `cause` to a non-error for the same reason I wouldn't `throw {non: 'error'}` and I'd hope a linter would agree with that.

To my dismay, MDN also includes using `cause` with non-errors :(

**Comment by andrei9669** [comment]
yeah, my only reference was the MDN, if only that didn't encourage such behavior. alright, I will rethink the logging in my project. I guess it's okay to close the issue.

**Comment by mdjermanovic** [comment]
Looks like we're not in favor of this change, so closing.
